pkg/controller/externalnode/controller.go

```diff
@@ -0,0 +1,383 @@
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in ExternalNode work queue but got %#v", obj)
+		return true
+	} else if err := c.syncExternalNode(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing ExternalNode", "ExternalNode", key)
+	}
+	return true
+}
+
+func (c *ExternalNodeController) syncExternalNode(key string) error {
+	namespace, name, err := splitKeyFunc(key)
+	if err != nil {
+		// This err should not occur.
+		return err
+	}
+	en, err := c.externalNodeLister.ExternalNodes(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return c.deleteExternalNode(namespace, name)
+	}
+
+	preEn, exists, _ := c.syncedExternalNode.GetByKey(key)
+	if !exists {
+		return c.addExternalNode(en)
+	} else {
+		return c.updateExternalNode(preEn.(*v1alpha1.ExternalNode), en)
+	}
+}
+
+func (c *ExternalNodeController) addExternalNode(en *v1alpha1.ExternalNode) error {
+	eeName := genExternalEntityName(en)
+	if eeName == "" {
+		klog.InfoS("Interfaces are empty for ExternalNode", "ExternalNode", klog.KObj(en))
+		return nil
+	}
+	ee := genExternalEntity(eeName, en)
+	err := c.addExternalEntity(ee)
+	if err != nil {
+		return err
+	}
+	c.syncedExternalNode.Add(en)
+	return nil
+}
+
+func (c *ExternalNodeController) addExternalEntity(ee *v1alpha2.ExternalEntity) error {
+	_, err := c.crdClient.CrdV1alpha2().ExternalEntities(ee.Namespace).Create(context.TODO(), ee, metav1.CreateOptions{})
+	if errors.IsAlreadyExists(err) {
+		klog.InfoS("Update ExternalEntity instead of creating it as it already exists", "ExternalEntity", klog.KObj(ee))
+		return c.updateExternalEntity(ee)
+	}
+	return err
+}
+
+func (c *ExternalNodeController) updateExternalNode(preEn *v1alpha1.ExternalNode, curEn *v1alpha1.ExternalNode) error {
+	if reflect.DeepEqual(preEn.Spec.Interfaces, curEn.Spec.Interfaces) && reflect.DeepEqual(preEn.Labels, curEn.Labels) {
+		return nil
+	}
+	// Delete the previous ExternalEntity and create a new one if the name of the generated ExternalEntity is changed.
+	// Otherwise, update the ExternalEntity.
+	if preEn.Spec.Interfaces[0].Name != curEn.Spec.Interfaces[0].Name {
+		preEEName := genExternalEntityName(preEn)
+		if preEEName != "" {
+			err := c.deleteExternalEntity(preEn.Namespace, preEEName)
+			if err != nil {
+				return err
+			}
+		}
+		curEEName := genExternalEntityName(curEn)
+		if curEEName != "" {
+			curEE := genExternalEntity(curEEName, curEn)
+			err := c.addExternalEntity(curEE)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		preIPs := sets.NewString(preEn.Spec.Interfaces[0].IPs...)
+		curIPs := sets.NewString(curEn.Spec.Interfaces[0].IPs...)
```

Comment:
ditto

Reply:
Thanks for pointing out. I have updated the function here to make sure len(en.Spec.Interfaces) > 0 here.